warp10/src/main/java/io/warp10/script/functions/ECRECOVER.java

```diff
@@ -163,9 +165,25 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
       z = z.shiftRight(hash.length * 8 - nbits);
     }
 
+    BigInteger rinv = spec.getCurve().fromBigInteger(r.modInverse(spec.getN())).toBigInteger();
+
     Set<Object> candidates = new HashSet<Object>();
 
-    for (int j = 0; j < H.intValue(); j++) {
+    int maxH = H.intValue();
+
+    if (maxH > MAX_COFACTOR) {
+      Map<String,String> capabilities = Capabilities.get(stack, (List) null);
+      if (null != capabilities.get(CAP_COFACTOR)) {
+        int hmax = Integer.valueOf(capabilities.get(CAP_COFACTOR));
+        if (maxH > hmax) {
+          throw new WarpScriptException(getName() + " cofactor " + maxH + " is above the maximum " + hmax + " allowed by the '" + CAP_COFACTOR + "' capability.");
+        }
+      } else {
+        throw new WarpScriptException(getName() + " cofactor " + maxH + " is above allowed maximum " + MAX_COFACTOR + ", increase this limit using a token with the '" + CAP_COFACTOR + "' capability.");
+      }
+    }
```

Comment:
Why don't you use `Capabilities.get(stack, CAP_COFACTOR);`?
Checking for a valid capability value is also nice and it can be merged with the null check like this:
```
    if (maxH > MAX_COFACTOR) {
      String hmaxCap = Capabilities.get(stack, CAP_COFACTOR);
      try {
        int hmax = Integer.valueOf(hmaxCap);
        if (maxH > hmax) {
          throw new WarpScriptException(getName() + " cofactor " + maxH + " is above the maximum " + hmax + " allowed by the '" + CAP_COFACTOR + "' capability.");
        }
      } catch (NumberFormatException nfe) {
        throw new WarpScriptException(getName() + " cofactor " + maxH + " is above allowed maximum " + MAX_COFACTOR + ", increase this limit using a token with the '" + CAP_COFACTOR + "' capability.");
      }
    }
```